pkg/apis/pipeline/v1/param_types.go

```diff
@@ -228,7 +221,7 @@ func (arrayOrString *ArrayOrString) applyOrCorrect(stringReplacements map[string
 
 	// if the stringVal is a string literal or a string that mixed with var references
 	// just do the normal string replacement
-	if !exactVariableSubstitutionRegex.MatchString(stringVal) {
+	if !isolatedResultVariableSubstitutionRegex.MatchString(stringVal) && !substitution.IsIsolatedArrayOrObjectParamRef(stringVal) {
```

Comment:
hm it seems like the `substitution` package is responsible for determining whether a reference is an isolated reference or not. Can the regex being used be moved to that package instead (and put into a more useful helper function instead of exporting the regex)? With the existing code, each package is only partially responsible for determining how the param reference should be parsed, and I'd like to see better separation of concerns.